### PR TITLE
Use .NET Core SequenceReader

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Remove BufferReader and use SequenceReader<T> which now ships in CoreFX.

This is related to https://github.com/aspnet/KestrelHttpServer/pull/3068 which builds on the functionality added to the reader.
I'll follow up with a port of said PR after this goes in.

Addresses #4693

cc: @danmosemsft, @davidfowl

### CODE REVIEW

**Review by pakrym — APPROVED** [review]

**Comment by pakrym** [comment]
We need to run a perf test with this to make sure we have no regressions.

cc @sebastienros

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs [hunk 1]
```diff
@@ -182,25 +182,26 @@ public unsafe bool ParseHeaders(TRequestHandler handler, in ReadOnlySequence<byt
...
                         while (remaining > 0)
                         {
-                            var index = reader.CurrentSegmentIndex;
-                            int ch1;
-                            int ch2;
+                            var index = reader.CurrentSpanIndex;
+                            byte ch1;
+                            byte ch2;
                             var readAhead = false;
+                            bool readSecond = true;
```
Comment (davidfowl) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs]: var

**Review by ahsonkhan — APPROVED** [review]

**Comment by natemcmaster** [comment]
Ok, i give up. The bot doesn't like us anymore.

**Comment by ahsonkhan** [comment]
> Ok, i give up. The bot doesn't like us anymore.

Let me give it a try, it always listens to me!

/AzurePipelines run ... away?

:)

**Comment by jkotalik** [comment]
I'm just going to make a separate temp PR for perf validation.

**Comment by jkotalik** [comment]
Results

| Description |       RPS | CPU (%) | Memory (MB) | Avg. Latency (ms) | Startup (ms) | First Request (ms) | Latency (ms) | Ratio |
| ----------- | --------- | ------- | ----------- | ----------------- | ------------ | ------------------ | ------------ | ----- |
|    Baseline | 2,165,914 |      91 |         385 |               1.8 |          212 |               36.6 |          0.5 |  1.00 |
|    SequenceReader | 2,172,325 |      93 |         181 |               2.1 |          224 |                 37 |          0.5 |  1.00 |

**Review by jkotalik — APPROVED** [review]

**Comment by halter73** [comment]
@jkotalik Should we rerun the benchmarks for this? @pakrym reminded me there was an issue with the benchmarks server last week.

**Comment by sebastienros** [comment]
not only an issue but the baseline is using a different runtime, look at the memory usage

**Comment by jkotalik** [comment]
Can you go ahead and rerun these with the correct runtimes then?

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by sebastienros** [comment]
I will do it

**Comment by sebastienros** [comment]
@JeremyKuhne would you mind merging master into your branch so I can run benchmarks?

**Comment by JeremyKuhne** [comment]
> @JeremyKuhne would you mind merging master into your branch so I can run benchmarks?

rebased

**Comment by halter73** [comment]
@sebastienros confirmed there isn't a regression.

**Comment by danmosemsft** [comment]
Yay.

**Pull request merged**